Source/HTTP/httpcall.cpp

```diff
@@ -137,6 +138,91 @@ HRESULT CALLBACK HC_CALL::PerfomAsyncProvider(XAsyncOp op, XAsyncProviderData co
     }
 }
 
+void CALLBACK HC_CALL::CompressRequestBody(void* c, bool canceled)
+{
+    PerformContext* context{ static_cast<PerformContext*>(c) };
+    HC_CALL* call{ context->call };
+
+    if (canceled)
+    {
+        XAsyncComplete(context->asyncBlock, E_ABORT, 0);
+        return;
+    }
+
+    if (call->traceCall)
+    {
+        HC_TRACE_INFORMATION(HTTPCLIENT, "HC_CALL::CompressRequestBody [ID %llu] Iteration %d", TO_ULL(call->id), call->m_iterationNumber);
+    }
+
+#if !HC_NOZLIB
+
+#if HC_PLATFORM == HC_PLATFORM_WIN32 || HC_PLATFORM == HC_PLATFORM_GDK
+    // Compress body before call if applicable
+    if (call->enableGzipCompression && call->compressionLevel != HCCompressionLevel::None)
+    {
+        HCHttpCallRequestBodyReadFunction clientRequestBodyReadCallback{ nullptr };
+        size_t requestBodySize{};
+        void* clientRequestBodyReadCallbackContext{ nullptr };
+        HRESULT hr = HCHttpCallRequestGetRequestBodyReadFunction(call, &clientRequestBodyReadCallback, &requestBodySize, &clientRequestBodyReadCallbackContext);
+
+        if (FAILED(hr) || !clientRequestBodyReadCallback)
+        {
+            HC_TRACE_ERROR(HTTPCLIENT, "CurlEasyRequest::ReadCallback: Unable to get client's RequestBodyRead callback");
+            return;
+        }
+        
+        size_t uncompressedRequestyBodySize = 0;
+        const size_t bufferSize = 128 * 1024;
+        uint8_t uncompressedRequestyBodyBuffer[bufferSize];
+
+        try
+        {
+            hr = clientRequestBodyReadCallback(call, 0, bufferSize, clientRequestBodyReadCallbackContext, uncompressedRequestyBodyBuffer, &uncompressedRequestyBodySize);
```

Comment:
>hr = clientRequestBodyReadCallback(call, 0, bufferSize, clientRequestBodyReadCallbackContext, uncompressedRequestyBodyBuffer, &uncompressedRequestyBodySize); [](http://example.com/codeflow?start=12&length=157)

What happens if the buffer isn't large enough here? I believe you'll need some logic to resize the buffer and call the client callback again with an offset

Reply:
I had a question about what the purpose of the offset was and this is actually answers that... so maybe adding like a loop where we check that we already got all the request bytes?

Reply:
Fixed, can you take a look at my solution?

Reply:
similar to my comment on the 'compress' method - I don't think you really need a second temp buffer here.  Use a vector + vector::resize as needed.  The reallocation and copying will be handled automatically that way

Reply:
Fixed

Reply:
I totally missed that we could query the full request body size up front with HCHttpCallRequestGetRequestBodyReadFunction.  In that case you can just create a vector large enough to hold the whole thing and avoid this loop altogether.  Sorry for making this more complicated than it needed to be!

Reply:
Fixed 🙂